src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -129,33 +267,437 @@ public ResultSet getSchemas() throws SQLException {
+        Pattern columnNamePatternRE = Pattern.compile(columnNamePattern);
+
+        Stream<BsonDocument> docs;
+        if (catalog == null) {
+            // If no catalog (database) is specified, get column privileges for all databases.
+            docs =
+                    this.conn
+                            .mongoClient
+                            .listDatabaseNames()
+                            .into(new ArrayList<>())
+                            .stream()
+                            .flatMap(
+                                    dbName ->
+                                            getColumnsFromDB(
+                                                    dbName,
+                                                    tableNamePatternRE,
+                                                    columnNamePatternRE,
+                                                    this::toGetColumnPrivilegesDoc));
+        } else {
+            docs =
+                    getColumnsFromDB(
+                            catalog,
+                            tableNamePatternRE,
+                            columnNamePatternRE,
+                            this::toGetColumnPrivilegesDoc);
+        }
+
+        // Per JDBC spec, sort by  COLUMN_NAME and PRIVILEGE. Since all PRIVILEGEs are the same,
+        // we just sort by COLUMN_NAME here.
+        List<BsonDocument> docsList =
+                docs.sorted(Comparator.comparing((BsonDocument doc) -> doc.getString(COLUMN_NAME)))
+                        .collect(Collectors.toList());
+
+        BsonExplicitCursor c = new BsonExplicitCursor(docsList);
+
+        return new MongoSQLResultSet(null, c, botSchema);
     }
 
     @Override
     public ResultSet getTablePrivileges(
             String catalog, String schemaPattern, String tableNamePattern) throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        MongoJsonSchema resultSchema = MongoJsonSchema.createEmptyObjectSchema();
+        resultSchema.addRequiredScalarKeys(
+                new Pair<>(TABLE_CAT, BSON_STRING_TYPE_NAME),
+                new Pair<>(TABLE_SCHEM, BSON_STRING_TYPE_NAME),
+                new Pair<>(TABLE_NAME, BSON_STRING_TYPE_NAME),
+                new Pair<>(GRANTOR, BSON_STRING_TYPE_NAME),
+                new Pair<>(GRANTEE, BSON_STRING_TYPE_NAME),
+                new Pair<>(PRIVILEGE, BSON_STRING_TYPE_NAME),
+                new Pair<>(IS_GRANTABLE, BSON_STRING_TYPE_NAME));
+
+        // All fields in this result set are nested under the bottom namespace.
+        MongoJsonSchema botSchema = MongoJsonSchema.createEmptyObjectSchema();
+        botSchema.properties.put(BOT_NAME, resultSchema);
+
+        // Note: JDBC has Catalogs, Schemas, and Tables: they are three levels of organization.
+        // MongoDB only has Databases (Catalogs) and Collections (Tables), so we ignore the
+        // schemaPattern argument.
+        Pattern tableNamePatternRE = Pattern.compile(tableNamePattern);
```

Comment:
Also looks like it is missing a toJavaPattern here.

Reply:
Done